File: bib.R
#!/usr/bin/Rscript --no-init-file

# definitions
indir = normalizePath("~/Dropbox/professional/archive") # dir with PDFs
reader = "/usr/bin/evince"                              # PDF reader
exiftool = "/usr/bin/exiftool"                          # check/set PDF title
file = "/usr/bin/file"                                  # check file mime type

# mirror
#mirror = "http://adsabs.harvard.edu"
mirror = "http://ukads.nottingham.ac.uk"
#mirror = "http://ads.nao.ac.jp"

# FUNCTION counter
counter = function(i, total, text = "", overline = FALSE, underline = FALSE, endnewline = TRUE){
    toprint = paste("  ", i, " / ", total, " ", sep="")
    if(text != ""){toprint = paste(toprint, ": ", text, " ", sep="")}
    linelen = nchar(toprint) + 1
    if(overline){
        toprint = paste(paste(rep("-",linelen), collapse=""), "\n", toprint, sep="")
        endnewline = FALSE
    }
    if(underline){
        toprint = paste(toprint, "\n", paste(rep("-",linelen), collapse=""), sep="")
        endnewline = FALSE
    }
    if(overline | underline){
        toprint = paste(toprint, "\n", sep="")
    }
    if(!overline & !underline){
        cat(paste(rep(c(" ","\b"),each=80),collapse=""))    # clear line
        todelete = nchar(toprint)
        backspace = paste(rep("\b", len=todelete), collapse="")
        toprint = paste(toprint, backspace, sep="")
    }
    if(endnewline){
        if(i == total){
            toprint = paste(toprint, "\n", sep="")
        }
    }
    cat(toprint)
}

# FUNCTION strip
strip = function(x, strip=" "){
    # workhorse function
    sfunc = function(x, strip){
        x = strsplit(x,"")[[1]]
        for(i in 1:length(strip)){
            s = strip[i]
            good = which(x!=s)
            if(length(good)>0){
                x = x[min(good):max(good)]
            }else{
                x = ""
            }
        }
        x = paste(x,collapse="")
        return(x)
    }
    # vectorize
    stripped = as.character(Vectorize(sfunc, vectorize.args=c("x"))(x=x, strip=strip))
    # return results
    return(stripped)
}

# split up stacked inputargs, define rds/bib files
inputargs = commandArgs(TRUE)
#inputargs = c("kelvin1899a", "-o")
if(length(inputargs)==0){inputargs = "-h"}
tempargs = as.list(inputargs)
char1 = substr(tempargs, 1, 1)
char2 = substr(tempargs, 2, 2)
tosplit = which(char1 == "-" & char2 != "-")
tempargs[tosplit] = lapply(X=strsplit(substr(inputargs[tosplit], 2, 50), ""), FUN=function(X){paste("-",X,sep="")})
inputargs = unlist(tempargs)
if(length(inputargs)>0){if(strsplit(inputargs[1], "")[[1]][1]=="-"){inputargs = c(" ",inputargs)}}
librds = paste(indir, "/bibtex.rds", sep="")
libbib = paste(indir, "/bibtex.bib", sep="")
libtxt = paste(indir, "/bibtex.txt", sep="")

# run argument(s)
if( (length(grep("-h",inputargs)) > 0)
    | (length(grep("--help",inputargs)) > 0)
    ){

    cat("
Usage: bib [arg] [INPUT]

Arguments:

bib -a,--all        (print full paper titles, no truncation)
bib -b,--bib        (include full bibtex entries in search)
bib -f,--fix        (run PDF title get/set routine over all PDFs)
bib -h,--help       (show help file)
bib -m,--max        (maximum number of entries to print)
bib -o,--open       (open first matching search entry in PDF reader)
bib -r,--reverse    (reverse order)
bib -u,--update     (update bibliography from ADS or local bibtex.txt file)
bib -w,--whole      (update bibliography from ADS or local and re-scan local)
bib -y,--year       (order by year)
")
    if(file.exists(librds)){cat("\nBibliography currently contains \033[1m",nrow(readRDS(librds)),"\033[0m entries.\n", sep="")}
    cat("\n")

}else if(
    (length(grep("-f",inputargs)) > 0)
    | (length(grep("--fix",inputargs)) > 0)
    ){

    # data
    files = grep(".pdf", dir(indir), value=TRUE)
    bits = matrix(unlist(strsplit(files, "::")),ncol=2,byrow=T)
    bits[,2] = unlist(strsplit(bits[,2], ".pdf"))

    # loop
    for(i in 1:length(files)){

        #  setup
        ref = bits[i,1]
        code = bits[i,2]
        counter(i, length(files), ref)
        myfile = paste0(indir, "/", files[i])
        filetype = system(paste0(file, ' --mime-type -b "', myfile, '"'), intern=TRUE)

        # check/set title only if PDF
        if(filetype == "application/pdf"){

            # title
            title.prelim = system(paste0(exiftool, ' -title "', myfile, '"'), intern=TRUE)
            if(length(title.prelim) > 0){
                title = strsplit(title.prelim, " : ")[[1]][2]
            }else{
                title = ""
            }
            if(is.na(title)){title = ""}

            # set title if necessary
            if(title != ref){
                system(paste0(exiftool, ' -title="', ref, '" -overwrite_original_in_place -preserve -quiet -m "', myfile, '"'))
            }

        }

    }

}else if(
    (length(grep("-u",inputargs)) > 0)
    | (length(grep("--update",inputargs)) > 0)
    | (length(grep("-w",inputargs)) > 0)
    | (length(grep("--whole",inputargs)) > 0)
    ){

    # data
    files = grep(".pdf", dir(indir), value=TRUE)
    bits = matrix(unlist(strsplit(files, "::")),ncol=2,byrow=T)
    bits[,2] = unlist(strsplit(bits[,2], ".pdf"))
    res = data.frame( ref = bits[,1]                    # reference
                    , year = rep(-9999,length(files))   # publication year
                    , title = character(length(files))  # title
                    , code = bits[,2]                   # ADS ID
                    , bibentry = I(vector("list",length(files))) # ADS bibtex entry
                    , stringsAsFactors = FALSE
                    )
    if(file.exists(librds)){
        old = readRDS(librds)
        bad = which(!old[,"ref"] %in% res[,"ref"])
        if(length(bad) > 0){old = old[-bad,]}
        pos = match(old[,"ref"], res[,"ref"])
        res[pos,] = old
    }
    if(file.exists(libtxt)){
        extra = readLines(libtxt)
    }else{
        extra = ""
    }

    # loop over each necessary entry
    if((length(grep("-w",inputargs)) > 0) | (length(grep("--whole",inputargs)) > 0)){
        lookup = which(res[,"year"]==-9999 | res[,"code"]=="bibtex.txt")
    }else{
        lookup = which(res[,"year"]==-9999)
    }
    for(i in lookup){

        # setup
        ref = res[i,"ref"]
        code = res[i,"code"]
        counter(which(lookup==i), length(lookup), ref)

        # ADS or local bibtex entry?
        if(code=="bibtex.txt"){

            bibnodes = c(grep("@", extra), (length(extra)+1))
            bibstart = grep(paste0("\\{",ref), extra)
            bibend = bibnodes[which(bibnodes > bibstart)[1]] - 1
            bibentry = extra[bibstart:bibend]
            if(bibentry[length(bibentry)] != ""){bibentry = c(bibentry, "")}
            year = as.numeric(strip(strsplit(strsplit(grep(" year", bibentry, v=T), "= ")[[1]][2], ",")[[1]][1], c("{","}")))
            title = strsplit(strsplit(grep(" title", bibentry, v=T), "\\{")[[1]][2], "\\}")[[1]][1]

        }else{

            # ADS URLs
            adsurl = paste(mirror, "/abs/", code, sep="")
            biburlold = paste(mirror, "/cgi-bin/nph-bib_query?bibcode=", code, "&data_type=BIBTEX&db_key=AST&nocookieset=1", sep="")
            biburl = paste(mirror, "/abs/", code, "/exportcitation", sep="")

            # get and parse BIBTEX from ADS
            bibfile = suppressWarnings(readLines(biburl, skipNul=TRUE))
            bibfile = gsub("&amp;", '&', bibfile)       # ampersand (impacts many below)
            bibfile = gsub("&#34;", '"', bibfile)       # quotation mark
            bibfile = gsub("\\\\&lt;", '<', bibfile)    # less than
            bibfile = gsub("\\\\&gt;", '>', bibfile)    # greater than
            bibfile = gsub("\\\\&le;", '<=', bibfile)   # less than or equal to
            bibfile = gsub("\\\\&ge;", '>=', bibfile)   # greater than or equal to

            # update reference
            firstline = suppressWarnings(grep('readonly=\"\">@',bibfile)[1])
            bibtype = tolower(strsplit(strsplit(bibfile[firstline], 'readonly=\"\">')[[1]][2], "\\{")[[1]][1])
            bibfile[firstline] = paste(bibtype, "{", ref, ",", sep="")
            finalline = suppressWarnings(grep('adsnote',bibfile)[1])+1

            # pick out metadata
            title = suppressWarnings(strip(strsplit(grep("title = ", bibfile, value=TRUE), "title =")[[1]][2], c(" ","{","}",",","\\",'"',"{","}")))
            year = suppressWarnings(as.numeric(strip(strsplit(grep("year =", bibfile, value=TRUE), "year =")[[1]][2], c(" ",","," ",'"'))))

            # generate bibtext entry
            bibentry = bibfile[firstline:finalline]
            if(bibentry[length(bibentry)] != ""){bibentry = c(bibentry,"")}

            # arXiv fix (MNRAS compatible)
            if(length(grep("e-prints", bibentry)) > 0){
                bibentry[grep("e-prints", bibentry)] = "  journal = {ArXiv e-prints},"
                vrow = grep("   adsurl = \\{", bibentry)
                bibentry = bibentry[c(1:vrow,vrow:length(bibentry))]
                bibentry[vrow] = '   volume = "{\\!\\!}",'
            }

            # replace %26 with ampersand (principally in HTML URLs)
            if(length(grep("%26", bibentry)) > 0){bibentry = gsub("%26", "&", bibentry)}

            # replace &#39; with '
            if(length(grep("&#39;", bibentry)) > 0){bibentry = gsub("&#39;", "'", bibentry)}

        }

        # add to global
        res[i,"year"] = year
        res[i,"title"] = title
        res[i,"bibentry"][[1]] = list(bibentry)

        # check/set PDF title, if PDF, and if appropriate
        myfile = paste0(indir, "/", files[i])
        filetype = system(paste0(file, ' --mime-type -b "', myfile, '"'), intern=TRUE)
        if(filetype == "application/pdf"){
            # title
            title.prelim = system(paste0(exiftool, ' -title "', myfile, '"'), intern=TRUE)
            if(length(title.prelim) > 0){
                title = strsplit(title.prelim, " : ")[[1]][2]
            }else{
                title = ""
            }
            if(is.na(title)){title = ""}
            # set title if necessary
            if(title != ref){
                system(paste0(exiftool, ' -title="', ref, '" -overwrite_original_in_place -preserve -quiet -m "', myfile, '"'))
            }
        }

    }

    # write results RDS file
    saveRDS(res, file=librds)

    # write  bibtex file
    bibinfo = unlist(res[,"bibentry"])
    cat(paste(bibinfo, "\n", sep=""), sep="", file=libbib)
    if(file.exists(librds)){cat("Bibliography currently contains \033[1m",nrow(readRDS(librds)),"\033[0m entries (\033[1m", nrow(res)-nrow(old), "\033[0m new).\n", sep="")}

}else{

    # data
    dat = readRDS(librds)
    refs = unlist(lapply(dat[,"ref"], formatC, mode="character", width=-14))
    numyears = as.numeric(dat[,"year"])
    years = unlist(lapply(as.character(numyears), formatC, mode="character", width=-4))
    titles = dat[,"title"]
    codes = unlist(lapply(dat[,"code"], formatC, mode="character", width=-19))
    nchars = nchar(refs) + nchar(codes)
    bibentries = unlist(lapply(dat[,"bibentry"], paste, collapse=" "))

    # sort by year
    if( (length(grep("-y",inputargs)) > 0)
        | (length(grep("--year",inputargs)) > 0)
        ){
        oo = order(numyears)
        refs = refs[oo]
        years = years[oo]
        titles = titles[oo]
        codes = codes[oo]
        bibentries = bibentries[oo]
    }

    # reverse ordering
    if( (length(grep("-r",inputargs)) > 0)
        | (length(grep("--reverse",inputargs)) > 0)
        ){
        oo = length(refs):1
        refs = refs[oo]
        years = years[oo]
        titles = titles[oo]
        codes = codes[oo]
        bibentries = bibentries[oo]
    }

    # maximum number of entries to print
    maxcount = 9999
    if( (length(grep("-m",inputargs)) > 0)
        | (length(grep("--max",inputargs)) > 0)
        ){
        varnum = c(grep("-m", inputargs) + 1, grep("--max", inputargs) + 1)[1]
        maxcount = as.numeric(inputargs[varnum])
    }

    # full length titles, trimmed to terminal width below (possibly incl. bibentries)
    if( (length(grep("-b",inputargs)) > 0)
        | (length(grep("--bib",inputargs)) > 0)
        ){
        out1 = paste("\n", paste(paste0("\033[1m", refs, "\033[0m"), titles, bibentries, sep=" ", collapse="\n"), "\n\n", sep="")
    }else{
        out1 = paste("\n", paste(paste0("\033[1m", refs, "\033[0m"), titles, sep=" ", collapse="\n"), "\n\n", sep="")
    }
    tmp1 = tempfile()
    cat(out1, file=tmp1)

    # print
    if( (length(grep("-a",inputargs)) == 0)
        & (length(grep("--all",inputargs)) == 0)
        ){
        system("tput rmam")
    }
    cat("\n")
    system(paste("cat ", tmp1, " | grep --max-count=", maxcount," --color=auto --ignore-case '", inputargs[1], "'", sep=""))
    cat("\n")
    system("tput smam")

    # clean up
    unlink(tmp1)

    # open file?
    if( (length(grep("-o",inputargs)) > 0)
        | (length(grep("--open",inputargs)) > 0)
        ){

        # write greppable tmp file without bolded font
        if( (length(grep("-b",inputargs)) > 0)
        | (length(grep("--bib",inputargs)) > 0)
        ){
            out2 = paste("\n", paste(refs, titles, bibentries, sep=" ", collapse="\n"), "\n\n", sep="")
        }else{
            out2 = paste("\n", paste(refs, titles, sep=" ", collapse="\n"), "\n\n", sep="")
        }
        tmp2 = tempfile()
        cat(out2, file=tmp2)

        # first matched file (for subsequent file opening)
        allmatches = as.character(suppressWarnings(system(paste("cat ", tmp2, " | grep --max-count=", maxcount," --color=never --ignore-case '", inputargs[1], "'", sep=""), intern=T)))
        if(length(allmatches) > 0){
            allmatchessplit = strsplit(allmatches, " +")
            allmatchessplit = allmatchessplit[lapply(allmatchessplit,length)>0]
            firstmatchref = allmatchessplit[[1]][1]
            firstmatchcode = codes[which(strip(refs) == firstmatchref)]
            firstmatchfile = paste0(indir, "/", firstmatchref, "::", firstmatchcode, ".pdf")
        }else{
            firstmatchfile = NULL
        }

        # open file
        if(length(firstmatchfile) > 0){
            comm = paste(reader, ' "', firstmatchfile, '" &', sep="")
            system(comm)
        }else{
            cat("ERROR: no file found\n")
        }

        # clean up
        unlink(tmp2)

    }

}

File: bib/bib.py
#!/usr/bin/env python3

import os
import sys
import tempfile
import pandas as pd
import urllib.request

pdfdir = os.path.expanduser('~/Dropbox/professional/archive')   # PDF dir
bibdir = os.path.expanduser('~/software/upsilon/bib')           # bibtex dir
reader = '/usr/bin/evince'                                      # PDF reader
exiftool = '/usr/bin/exiftool'                                  # mod PDF title
file = '/usr/bin/file'                                          # mime types
bibfile = bibdir + '/bibtex.bib'                                # bibtex file

mirror = "http://adsabs.harvard.edu"
# mirror = "http://ukads.nottingham.ac.uk"
# mirror = "http://ads.nao.ac.jp"

bld = '\033[1m'
nrm = '\033[0m'

raw_args = sys.argv[1:]
args = []
for arg in raw_args:
    if arg[0] == '-':
        args.extend([f'-{x}' for x in arg[1:]])
    else:
        args.append(arg)
args.sort(reverse=True)

# usage when rebuilding bib file from scratch
# CL install privoxy tor
# CL add 'forward-socks4a / localhost:9050' to '/etc/privoxy/config'
# CL service tor start
# CL service privoxy start
# CL pip3 install pysocks
# import socks
# import socket
# socks.set_default_proxy(socks.SOCKS5, "localhost", 9050)
# socket.socket = socks.socksocket
# CL service tor stop
# CL service privoxy stop


def bibread(bibfile):
    with open(bibfile) as f:
        bibdat = [x for x in f.readlines() if x != '\n']
    if bibdat[-1][-1:] != '\n':
        bibdat[-1] += '\n'
    rows = [i for i, x in enumerate(bibdat) if '@' in x] + [len(bibdat)+1]
    refs, years, titles, bibentries = [], [], [], []
    for first, last in zip(rows[:-1], rows[1:]):
        bibentry = ''.join(bibdat[first:last])
        ref = bibentry.split('{')[1].split('\n')[0].strip(',"{}')
        year = int(bibentry.split('year = ')[1].split('\n')[0].strip(',"{}'))
        title = bibentry.split('title = ')[1].split('\n')[0].strip(',"{}')
        refs.append(ref)
        years.append(year)
        titles.append(title)
        bibentries.append(bibentry)
    df = pd.DataFrame(dict(ref=refs, year=years, title=titles,
                           bibentry=bibentries))
    # abc ref sort disabled here for speed, enabled for writing instead
    # df = df.iloc[df.ref.str.lower().argsort()]
    # df.reset_index(inplace=True, drop=True)
    return df


def bibwrite(df, bibfile):
    df = df.iloc[df.ref.str.lower().argsort()]
    df.reset_index(inplace=True, drop=True)
    with open(bibfile, 'w') as f:
        for bibentry in df['bibentry']:
            f.write(bibentry + '\n')


try:
    df = bibread(bibfile)
except Exception:
    raise


if len(args) == 0 or '-h' in args:
    print('')
    print('Usage: bib [INPUT] [arg]')
    print('')
    print('Arguments:')
    print('')
    print('bib -a   (print full paper titles, no truncation)')
    print('bib -b   (include full bibtex entries in search)')
    print('bib -f   (run PDF title get/set routine over all PDFs)')
    print('bib -h   (show help file)')
    print('bib -o   (open first matching search entry in PDF reader)')
    print('bib -r   (reverse order)')
    print('bib -u   (update bibliography from ADS / local bibtex.txt file)')
    print('bib -y   (order by year)')
    print('')
    print(f'Bibliography contains {bld}{len(df)}{nrm} entries.')
    print('')
elif not os.path.exists(pdfdir) and ('-u' in args or '-f' in args
                                     or '-o' in args):
    raise FileNotFoundError(f'Cannot locate PDF directory: {pdfdir}')
elif '-u' in args:
    all_files = os.listdir(pdfdir)
    pdfs = sorted([x for x in all_files if '.pdf' in x])
    refs, pdf_codes = zip(*[x.split('::') for x in pdfs])
    codes = [x[:-4] for x in pdf_codes]
    findreplace = {'&amp;': '&',
                   '&#34;': '"',
                   '\\\\&lt;': '<',
                   '\\\\&gt;': '>',
                   '\\\\&le;': '<=',
                   '\\\\&ge;': '>=',
                   '%26': '&',
                   '&#39;': "'"}

    df2 = pd.DataFrame(dict(ref=refs,
                            year=[-9999]*len(refs),
                            title=['']*len(refs),
                            bibentry=['']*len(refs),
                            code=codes,))
    df2 = df2.iloc[df2.ref.str.lower().argsort()]
    df2.reset_index(inplace=True, drop=True)
    for index, row in df2.iterrows():
        if row['ref'] in df['ref'].to_list():
            old_index = df.index[df['ref'] == row['ref']].tolist()[0]
            df2.at[index, 'year'] = df.at[old_index, 'year']
            df2.at[index, 'title'] = df.at[old_index, 'title']
            df2.at[index, 'bibentry'] = df.at[old_index, 'bibentry']

    for index, row in df2.iterrows():
        print(f'{index} : {row.ref} ', end='', flush=True)
        if row.year == -9999 and row.code != 'LSK':
            biburl = mirror + '/abs/' + row.code + '/exportcitation'
            bibresponse = urllib.request.urlopen(biburl)
            bibhttp = []
            for line in bibresponse:
                bibhttp.append(line.decode("utf-8"))
            l0 = [i for i, x in enumerate(bibhttp)
                  if 'readonly=\"\">@' in x][0] + 1
            l1 = [i for i, x in enumerate(bibhttp)
                  if 'adsnote' in x][0] + 2
            bibtype = bibhttp[l0 - 1].split('"">@')[1].split('{')[0]
            bibref = '@'+bibtype.upper()+'{'+row.ref+',\n'
            bibentry = ''.join(bibhttp[l0:l1])
            for key in findreplace:
                bibentry = bibentry.replace(key, findreplace[key])
            if 'BOOK' in bibtype and len(bibentry.split('publisher =')) == 1:
                publisher = input('publisher: ')
                bibref += '    publisher = {' + publisher + '},\n'
            if 'BOOK' in bibtype and len(bibentry.split('editor =')) == 2:
                bibentry = bibentry.replace('editor =', '  note =')
            temp_year = bibentry.split('year = ')[1].split('\n')[0]
            temp_title = bibentry.split('title = ')[1].split('\n')[0]
            df2.at[index, 'year'] = int(temp_year.strip(',"{}'))
            df2.at[index, 'title'] = temp_title.strip(',"{}')
            df2.at[index, 'bibentry'] = bibref + bibentry
            df2 = df2.astype({'year': int})
            pdfpath = f'{pdfdir}/{row.ref}::{row.code}.pdf'
            exifsetcmd = (f'{exiftool} -title="{row.ref}" '
                          f'-overwrite_original_in_place '
                          f'-preserve -quiet -m "{pdfpath}"')
            titleset = os.popen(exifsetcmd).read().strip('\n')
        print('\x1b[2K', end='\r')

    bibwrite(df2, bibfile)
    print((f'Bibliography contains {bld}{len(df2)}{nrm} entries '
           f'({bld}{len(df2) - len(df):+d}{nrm} change)'))
elif '-f' in args:
    all_files = os.listdir(pdfdir)
    pdfs = sorted([x for x in all_files if '.pdf' in x])
    refs, pdf_codes = zip(*[x.split('::') for x in pdfs])

    for index, (ref, pdf) in enumerate(zip(refs, pdfs)):
        print(f'{index} : {ref} ', end='', flush=True)
        pdfpath = pdfdir + '/' + pdf
        filetype = os.popen(f'{file} --mime-type -b "{pdfpath}"').read()
        if 'application/pdf' in filetype:
            titleline = os.popen(f'{exiftool} -title "{pdfpath}"').read()
            if ref + '\n' not in titleline:
                exifsetcmd = (f'{exiftool} -title="{ref}" '
                              f'-overwrite_original_in_place '
                              f'-preserve -quiet -m "{pdfpath}"')
                titleset = os.popen(exifsetcmd).read().strip('\n')
        print('\x1b[2K', end='\r')
else:
    if '-y' in args:
        df = df.iloc[df.year.argsort()]
        # df.reset_index(inplace=True, drop=True)
    if '-r' in args:
        df = df.iloc[::-1]
        # df.reset_index(inplace=True, drop=True)

    if '-a' not in args:
        linetruncon = os.system('tput rmam')
    tmp = tempfile.NamedTemporaryFile(mode='w+')
    for ref, tit, be in zip(df.ref, df.title, df.bibentry):
        # extra spaces required at end of line before \n to enable grepping
        # not sure why this is the case, but issue verified during testing
        if '-b' in args:
            be = be.replace('\n', '')
            tmp.write(f'{bld}{ref:14s}{nrm} {tit} {be}  \n\n')
        else:
            tmp.write(f'{bld}{ref:14s}{nrm} {tit}  \n')
    catcmd = f'grep --color=auto --ignore-case "{args[0]}" {tmp.name}'
    if '-a' in args and '-b' in args:
        catcmd += ' -A 1'
    print('')
    catstat = os.system(catcmd)
    if catstat == 0 and '-a' not in args and '-b' not in args:
        print('')
    if '-o' in args:
        catout = os.popen(catcmd).read()
        matchref = catout.split(bld)[1].split(nrm)[0].strip()
        all_files = os.listdir(pdfdir)
        pdfs = sorted([x for x in all_files if '.pdf' in x])
        matchpdf = pdfdir + '/' + [x for x in pdfs if matchref+'::' in x][0]
        readercmd = f'{reader} "{matchpdf}" &'
        readerout = os.system(readercmd)
    tmp.close()
    linetruncoff = os.system('tput smam')
